utils/merkletrie/fsnoder/dir.go

```diff
@@ -0,0 +1,131 @@
+package fsnoder
+
+import (
+	"bytes"
+	"fmt"
+	"hash/fnv"
+	"sort"
+
+	"gopkg.in/src-d/go-git.v4/utils/merkletrie/noder"
+)
+
+// Dir values implement directory-like noders.
+type dir struct {
+	name     string        // relative
+	children []noder.Noder // sorted by name
+	hash     []byte        // memoized
+}
+
+// copies the children slice, so nobody can modify the order of its
+// elements from the outside.
+func newDir(name string, children []noder.Noder) (*dir, error) {
+	cloned := make([]noder.Noder, len(children))
+	_ = copy(cloned, children)
+	sort.Sort(noder.ByName(cloned))
+
+	if hasChildrenWithNoName(cloned) {
+		return nil, fmt.Errorf("non-root inner nodes cannot have empty names")
+	}
+
+	if hasDuplicatedNames(cloned) {
+		return nil, fmt.Errorf("children cannot have duplicated names")
+	}
+
+	return &dir{
+		name:     name,
+		children: cloned,
+	}, nil
+}
+
+func hasChildrenWithNoName(children []noder.Noder) bool {
+	for _, c := range children {
+		if c.Name() == "" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func hasDuplicatedNames(children []noder.Noder) bool {
+	if len(children) < 2 {
+		return false
+	}
+
+	for i := 1; i < len(children); i++ {
+		if children[i].Name() == children[i-1].Name() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (d *dir) Hash() []byte {
+	if d.hash == nil {
+		d.calculateHash()
+	}
+
+	return d.hash
+}
+
+// hash is calculated as the hash of "dir " plus the concatenation, for
+// each child, of its name, a space and its hash.  Children are sorted
+// alphabetically before calculating the hash, so the result is unique.
+func (d *dir) calculateHash() {
+	h := fnv.New64a()
+	h.Write([]byte("dir "))
+	for _, c := range d.children {
+		h.Write([]byte(c.Name()))
+		h.Write([]byte(" "))
+		h.Write(c.Hash())
+	}
+	d.hash = h.Sum([]byte{})
+}
+
+func (d *dir) Name() string {
+	return d.name
+}
+
+func (d *dir) IsDir() bool {
+	return true
+}
+
+// returns a copy so nobody can alter the order of its elements from the
+// outside.
+func (d *dir) Children() ([]noder.Noder, error) {
```

Comment:
We don't need it, but tests are easier to build if this is true, and it does not affect the performance of the production code, as the `fsnoder` package should only be used in tests.